### PR TITLE
HRCPP-499 configurable act on pool exhausted and test

### PR DESCRIPTION
**Pull request by rigazilla**
This allows the user to configure the recover policy when connection pool is exhausted:

    configurationBuilder.connectionPool().exhaustedAction(WAIT)
waits indefinitely for a connection to be returned and made available

    configurationBuilder.connectionPool().exhaustedAction(EXCEPTION)
raises a NoSuchElementException

### CODE REVIEW

File: test/PoolUnitTest.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		if (flag == 1) {
+			cp.borrowObject(addr);
+			break;
+		}
+	}
+	t1.join();
+	std::cout << "OK" << std::endl;
+}
+
+/** testExec tests the pool in EXECEPTION configuration
```
Comment (alanfx) [inline test/PoolUnitTest.cpp]: nit: comment says `testExec` and the method is `testExc`

File: test/PoolTest.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		// Start 50 threads and see if everything works.
+		// With the EXCEPTION setup the user code must implement
+		// a recover policy if pool is exhausted, this test just waits
+		// for 1 sec and retry
+		std::string *res[50];
+		std::thread t[50];
+		std::promise<void> p[50];
+		int exceptionsCaught = 0;
+		for (int i = 0; i < 50; i++) {
+			t[i] =
```
Comment (alanfx) [inline test/PoolTest.cpp]: The formatting here looks a little weird?

File: src/hotrod/impl/transport/tcp/ConnectionPool.cpp [hunk 3]
```diff
@@ -124,7 +124,12 @@ TcpTransport& ConnectionPool::borrowObject(const InetSocketAddress& key) {
 			obj = &factory->makeObject(key); // then create new object
 		} else {
 			sys::ScopedUnlock<sys::Mutex> u(lock);
-			obj = idleQ->pop();  			 // else wait for the first available
+	        if (this->getConfiguration().getExhaustedAction() == EXCEPTION ) {
+				obj = idleQ->popOrThrow();       // else wait for the first available
```
Comment (alanfx) [inline src/hotrod/impl/transport/tcp/ConnectionPool.cpp]: This formatting also looks strange

File: test/PoolUnitTest.cpp (hunk 1 above)
Comment (rigazilla) [reply test/PoolUnitTest.cpp]: changed func name

File: test/PoolTest.cpp (hunk 2 above)
Comment (rigazilla) [reply test/PoolTest.cpp]: reformatted

File: src/hotrod/impl/transport/tcp/ConnectionPool.cpp (hunk 3 above)
Comment (rigazilla) [reply src/hotrod/impl/transport/tcp/ConnectionPool.cpp]: reformatted

**Comment by alanfx** [comment]
Looks good, @rigazilla Merging

**Pull request merged**